# Probability: Introduction to Foundational Concepts


## Objectives

- Describe set theory and its terminology
- Recognize combinatorics
- Describe the fundamentals of probability theory
- Define conditional probability


![xkcd](images/increased_risk_2x.png)
[[Image Source]](https://xkcd.com/1252/)

# Set Theory Basics

In probability theory, a **set** is a well-defined collection of objects (called **elements**). An element is either in a given set or not, and order is not considered.

> If an **element** $x$ belongs to a set $S$, then you'd write $x \in S$. 
> 
> On the other hand, if $x$ does not belong to $S$, then you'd write $x\notin S$.

Although sets can be abstract mathematical object, we'll typically think of sets as a collection of things sharing some relevant attribute(s).

## Set Notation

<img src="images/new_venn_diagram.png" alt="venn diagram with set notation" width=500>

The **union** of 2 sets $A$ and $B$ is the set $T$ of elements in either $A$ or $B$, or in both. We denote this with the symbol $\cup$.


$$\large T = A\cup B = B\cup A$$

The **intersection** of two sets $A$ and $B$ is the set $V$ that contains all elements of $A$ that also belong to $B$. We denote this with the symbol $\cap$.

$$\large V = A\cap B = B\cap A$$

<img src="images/setnotation.jpg" alt="set notation, found from https://slideplayer.com/slide/10502152/" width=800>

[Image Source](https://slideplayer.com/slide/10502152/)

More on Sets: https://www.mathsisfun.com/sets/sets-introduction.html

### But Now, in Python:

 These are set methods - work only on Python sets!

| Method                      | Result |
| ------                      | ------ |
| `s.issubset(t)`             | test whether every element in s is in t |
| `s.issuperset(t)`           | test whether every element in t is in s |
| `s.union(t)`                | new set with elements from both s and t |
| `s.intersection(t)`         | new set with elements common to s and t |
| `s.difference(t)`           | new set with elements in s but not in t |
| `s.symmetric_difference(t)` | new set with elements in either s or t but not both |

# Combinatorics

Before we dive in much further, let's briefly talk about how you can take elements from different collections and group them.

### Combinatorics Definition

> Combinatorics is the branch of mathematics that deals with the relations characterizing sets, subsets, lists, and multisets.
>
> Sometimes combinatorics is said to be the branch of math that deals with counting; and that’s true, but not in the sense in which you learned to count in kindergarten. Though combinatorics deals with numbering and finding out how many members are in sets, it’s designed to find ways of doing that **without actual, potentially tedious, counting** involved.

-- [Statistics How To](https://www.statisticshowto.com/probability-and-statistics/combinatorics/)

### What's the difference between a permutation and a combination?

In a **permutation**, order matters. If you have a race, it matters who arrives in first, or second, or third place - there's a difference in the ordering of the group!

> The number of **permutations** of **n** objects taken **r** at a time is given by the formula:
>
> $$\large P(n,r) = \frac{n!}{(n - r)!}$$

In a **combination**, you only care about which items are members of the set. For example, if you're creating groups of students to work on a project, the order in which you add their names to the group doesn't really matter - it's the group members itself, not any order, that you care about.

> The number of **combinations** of **n** objects taken **r** at a time is given by the formula:
>
> $$\large C(n,r) = \frac{n!}{r!(n - r)!}$$

Main things to ask when dealing with combinations or permutations:
- Does order matter? 
- With or without replacement? (aka does the first choice then restrict the second choice?)

### Q: How many possible orders are there for first/second/third place in a race with 30 contestants?

In [9]:
import math

#How many possible orders are there for first/second/third place in a race with 30 contestants?
#a 30 factorial = 30*29*28*27*......1 
#30 total contestants curious about 3 locatiosn
method_1 = math.factorial(30) / math.factorial(30-3)
method_1_1 = math.perm(30, 3)
method_1, method_1_1

(24360.0, 24360)

In [12]:
# Also the python library itertools has combination and permutations!
import itertools

#How many variations of the top 3 runners with 30 contestants 
#1 through 31;;30 options for top 3 // list 
#allows you iterate over a potentional outcomes and generates variations
method_2 = list(itertools.permutations(range(1,31),3))

method_2

[(1, 2, 3),
 (1, 2, 4),
 (1, 2, 5),
 (1, 2, 6),
 (1, 2, 7),
 (1, 2, 8),
 (1, 2, 9),
 (1, 2, 10),
 (1, 2, 11),
 (1, 2, 12),
 (1, 2, 13),
 (1, 2, 14),
 (1, 2, 15),
 (1, 2, 16),
 (1, 2, 17),
 (1, 2, 18),
 (1, 2, 19),
 (1, 2, 20),
 (1, 2, 21),
 (1, 2, 22),
 (1, 2, 23),
 (1, 2, 24),
 (1, 2, 25),
 (1, 2, 26),
 (1, 2, 27),
 (1, 2, 28),
 (1, 2, 29),
 (1, 2, 30),
 (1, 3, 2),
 (1, 3, 4),
 (1, 3, 5),
 (1, 3, 6),
 (1, 3, 7),
 (1, 3, 8),
 (1, 3, 9),
 (1, 3, 10),
 (1, 3, 11),
 (1, 3, 12),
 (1, 3, 13),
 (1, 3, 14),
 (1, 3, 15),
 (1, 3, 16),
 (1, 3, 17),
 (1, 3, 18),
 (1, 3, 19),
 (1, 3, 20),
 (1, 3, 21),
 (1, 3, 22),
 (1, 3, 23),
 (1, 3, 24),
 (1, 3, 25),
 (1, 3, 26),
 (1, 3, 27),
 (1, 3, 28),
 (1, 3, 29),
 (1, 3, 30),
 (1, 4, 2),
 (1, 4, 3),
 (1, 4, 5),
 (1, 4, 6),
 (1, 4, 7),
 (1, 4, 8),
 (1, 4, 9),
 (1, 4, 10),
 (1, 4, 11),
 (1, 4, 12),
 (1, 4, 13),
 (1, 4, 14),
 (1, 4, 15),
 (1, 4, 16),
 (1, 4, 17),
 (1, 4, 18),
 (1, 4, 19),
 (1, 4, 20),
 (1, 4, 21),
 (1, 4, 22),
 (1, 4, 23),
 (1, 4, 24),
 (1, 4,

#### * Can you spot the difference b/t methods?

In [14]:
#itertools will give you specific counts and if only interested in the number use math modules
#math doesn't allow replacemnt, itertool will

itertools.permutations(40,3)

TypeError: 'int' object is not iterable

-

### Q: How many possible codes are there for a standard padlock?

> Hint: (there are 40 numbers on a padlock and use 3 numbers - **with replacement**)

In [18]:
#40 potentional choices, #3 combinations #list gives the amount of solutions #length gives us the count of solutions
num=list(range(1,41))
#allows us to multiply itertools andthe list allows to catalog permutations
len(list(itertools.product(num, repeat =3)))
#40*40*40

64000

In [20]:
math.perm(40,3)

59280

For the first number: 40 choices. For the second number, still 40 choices: $40\cdot40=40^2$. Again, for 3rd number, still 40 choices: $40\cdot40\cdot40=40^3$

### Q: How many unique 3 topping pizzas can you make from the following 8 ingredients:

- Mushrooms
- Pepperoni
- Onion
- Peppers
- Ham
- Pineapple
- Sausage
- Olives
    
> Side note: which is the worst?

In [22]:
# Using math

three_top_pizzas = math.factorial(8)/(math.factorial(3)*math.factorial
              

SyntaxError: unexpected EOF while parsing (<ipython-input-22-85ceed781c1a>, line 4)

In [23]:
# Our list of toppings
toppings = ["Mushrooms", "Pepperoni", "Onion", "Peppers", 
            "Ham", "Pineapple", "Sausage", "Olives"]

# Using itertools
three_topping_pizzas = None
len(three_topping_pizzas)

TypeError: object of type 'NoneType' has no len()

# Probability Theory

Probability theory is the study of the frequency of a given event occurring with respect to all possible events.

**Probability is the likelihood of a specific outcome occurring out of all possible outcomes, expressed as a fraction between 0 and 1.**

But perhaps more importantly:

> **"Probabilities do not tell us what will happen for sure; they tell us what is _likely to happen_ and what is _less likely to happen_."**
>
> -- _Naked Statistics_, by Charles Wheelan, p. 72

## Why Should I Care About Probabilities?

Studying probabilities allows us to make better and more informed decisions, based on data previously collected. For example, understanding the fact that it is nearly impossible for us to ever win the lottery from a probabilistic standpoint deters us from using it as a source of income.

Probability theory also lies at the heart of making inferences using our data, which is what statistics is all about!

## Probability Terminology

From [_Introduction to Probability, Statistics, and Random Processes_](https://www.probabilitycourse.com/chapter1/1_3_1_random_experiments.php):

> **Outcome:** A result of a random experiment.
>
> **Sample Space:** The set of all possible outcomes.
>
> **Event:** A subset of the sample space.

To be more specific, an **event** is a specific outcome of a random experiment, often what we are hoping to predict (so it's sometimes called a successful outcome)

We typically write probability statements using notation like $P(E) = 0.5$, where $E$ is an event, such as getting Heads in a coin flip.

## Probability Using Sets

In its most basic form, the **probability** of an event occurring is the size of the specific outcome set divided by the size of the sample space. In some ways, probability is just about counting outcomes.

**Example**: What is the probability of rolling an even number on a six-sided die?

* Event: {2, 4, 6} has size 3
* Sample space: {1, 2, 3, 4, 5, 6} has size 6
* Thus the probability = 3/6 = 1/2 = 50%

#### Relevant Distinction: Types of Variables

A random variable can either be continuous or discrete
- **Continuous**: the variable can have any values within a range (e.g. a person's height)
- **Discrete**: the variable can only have a finite subset of values within a range (e.g. number of cars owned)

This is important because, for continuous variables, the probability of some specific event x _will functionally be zero_! BUT "in reality we are always interested in the probability of some intervals rather than a specific point x" [[Source]](https://www.probabilitycourse.com/chapter1/1_3_5_continuous_models.php)

## Calculating Probabilities

Note: some of this text comes from [_OpenIntro Statistics_](https://www.openintro.org/book/os/), Chapter 3.

In general, you can think of dividing the outcome you're exploring by all possible outcomes:

$$\large P(Event) = \frac{|Event|}{|Sample\ Space|} $$

#### General Addition Rule

The probability that either $A$ or $B$ will occur can be calculated by adding each individual probability, and then subtracting the probability that both occur together:

$$\large P(A \cup B) = P(A) + P(B) − P(A \cap B)$$

Remember, $P(A \cap B)$ expresses the overlap between the two events - if you don't subtract that overlap, then you double count the instances when **both** $A$ and $B$ occur!

#### Multiplication Rule for Independent Processes

A special condition is when the outcome of $A$ has no bearing on the outcome of $B$. We say these two events are **independent** (e.g. rolling a die and tossing a coin).

If $A$ and $B$ represent events from two different and **independent** processes, then the probability
that both $A$ and $B$ occur can be calculated as the product of their separate probabilities:

$$\large P(A \cap B) = P(A) * P(B)$$

### 🧠 Knowledge Check

### 1) AND Question:

What is the probability of rolling a 5 on a fair die _and_ getting a tails on a fair coin toss?

**Your answer here**:

- 


In [25]:
#What is the probability of rolling a 5 on a fair die and getting a tails on a fair coin toss?
(1/6) * (1/2)
#if independent use multicpiation 

0.08333333333333333


<details>
    <summary>Answer</summary>

We're checking for the intersection of these sets. Of the six possible outcomes on a die roll only one (the 5) will do. So the chance of getting a 5 on a die is 1/6. Of the two possible outcomes on a coin toss again only one (tails) will do. So the chance of getting tail on a coin toss is 1/2.

So the calculation is: $$\large P(5 \cap tails) = \left(\frac{1}{6}\right)*\left(\frac{1}{2}\right) = \frac{1}{12}$$
</details>

### 2) OR Question: 

What is the probability of rolling a 5 on a die _or_ getting a tails on a coin toss?

**Your answer here**:

- 


In [28]:
#What is the probability of rolling a 5 on a die or getting a tails on a coin toss?
(1/6) + (1/2) - (1/12)
#if depedent use + / - 

0.5833333333333333


<details>
    <summary>Answer</summary>
    
   We're now checking for the union of these sets. Here we want to count all the die-coin combinations where we have a 5 on the die AND all the die-coin combinations where we have a tails on the coin. 

$$\large P(5 \cup tails) $$

**BUT:**

If the die is 5, that includes two possibilities: 5-heads and **5-tails**.

Our coin is tails, that includes six possibilities: 1-tails, 2-tails, 3-tails, 4-tails, **5-tails**, and 6-tails.

But then we've counted the combination where **both** the 5 and the tails occur **twice**.

So the correct calculation is the sum of the individual probabilities **less the probability of their intersection**:

$$\large P(5 \cup tails) = \left(\frac{1}{6}\right) + \left(\frac{1}{2}\right) - \left(\left(\frac{1}{6}\right)*\left(\frac{1}{2}\right)\right) = \frac{7}{12} $$
    </details>

## Enough Talk - Let's Explore in Python!

### Mushroom dataset

Let's look at a modified version of the Mushroom dataset from UCI [here](https://archive.ics.uci.edu/ml/datasets/Mushroom). Each row in this dataset corresponds to one observation (one mushroom). 

In [29]:
import pandas as pd

In [30]:
df = pd.read_csv('data/Mushrooms_cleaned.csv')
df.head()

edible-poisonous gill-spacing stalk-shape stalk-color-above-ring  \
0        poisonous        close   enlarging                  white   
1           edible        close   enlarging                  white   
2           edible        close   enlarging                  white   
3        poisonous        close   enlarging                  white   
4           edible      crowded    tapering                  white   

  stalk-color-below-ring gill-color  bruised  
0                  white      black     True  
1                  white      black     True  
2                  white      brown     True  
3                  white      brown     True  
4                  white      black    False

In [31]:
df.describe()

edible-poisonous gill-spacing stalk-shape stalk-color-above-ring  \
count              8124         8124        8124                   8124   
unique                2            2           2                      9   
top              edible        close    tapering                  white   
freq               4208         6812        4608                   4464   

       stalk-color-below-ring gill-color bruised  
count                    8124       8124    8124  
unique                      9         12       2  
top                     white       buff   False  
freq                     4384       1728    4748

#### 1) If you picked a row from this dataset at random, what is the probability it corresponds to a bruised mushroom? 

In other words, find $P(bruised)$

In [35]:
#) If you picked a row from this dataset at random, what is the probability 
#it corresponds to a bruised mushroom?
#looking for an individusal qualifer 
event_space =len(df.loc[df['bruised'] == True])
#dividing that by the total space
sample_space = len(df)

In [37]:
p_bruised = event_space/sample_space
p_bruised

0.4155588380108321

In [39]:
# Another way
df['bruised'].value_counts(normalize=True)

False    0.584441
True     0.415559
Name: bruised, dtype: float64

In [ ]:
# Let's see...
df.sample(1)

#### 2) What is the probability you pick a row corresponding to a mushroom that is bruised _AND_ edible?

$P(edible \cap bruised)$

BUT! Are they independent events ...?


In [40]:
##### 2) What is the probability you pick a row corresponding
#to a mushroom that is bruised _AND_ edible?

p_bruised_and_edible = df[(df['bruised'] == True) & (df['edible-poisonous'] == 'edible')].shape[0]/df.shape[0]

In [41]:
p_bruised_and_edible

0.33874938453963566

Are being bruised and being edible independent of each other?

> Formally, $A$ and $B$ are *independent* if and only if the probability that *both* $A$ *and* $B$ happen is:
> 
> $$\large P(A \cap B) = P(A) * P(B)$$

In [42]:
p_bruised

0.4155588380108321

In [43]:
p_edible = len(df[df['edible-poisonous'] == 'edible']) / len(df)
p_edible

0.517971442639094

In [44]:
p_bruised * p_edible

0.21524761082589627

In [45]:
p_bruised_and_edible == p_bruised * p_edible

False

## Enter: Conditional Probability

### When do we compute conditional probabilities? 

We need to compute conditional probabilities when the outcome of an event depends on the outcome of previous events (**dependent** events). A conditional probability of an event is the probability of the event *given* another event has occurred.


When events _are_ independent, the rule for probabilistic AND (I'll use '$\cap$' below) is simple:

$$\large P(A\cap B) = P(A) * P(B)$$

But the more general rule, which includes non-independent events, is:

$$\large P(A\cap B) = P(A | B) * P(B)$$

> (this is pretty much the law of total probability, we'll revisit this later!)

In fact, this is the definition of conditional probability. Rearranging:

$$\large P(A | B) = \frac{P(A\cap B)}{P(B)}$$

The `|` here should be read as "given". We are given some information, $B$, and thus it reduces our sample space!

### Back to our Mushrooms!

#### 3) What is the probability of picking an edible mushroom given it is bruised? 

$P(edible | bruised)$

In [46]:
# What is the probability of picking an edible mushroom given it is bruised? 

p_edible_given_bruised = p_bruised_and_edible / p_bruised
p_edible_given_bruised

0.8151658767772512

In [48]:
# Same as checking using value_counts
bruised = df[df['bruised'] == True] # just bruised mushrooms
bruised['edible-poisonous'].value_counts(normalize=True)

edible       0.815166
poisonous    0.184834
Name: edible-poisonous, dtype: float64

#### 4) What is the probability of picking a bruised mushroom given it is edible? 

$P(bruised | edible)$

In [50]:
# What is the probability of picking a bruised mushroom given it is edible? 
p_edible = df[df['edible-poisonous'] == 'edible' ].shape[0]/df.shape[0]
p_edible

0.517971442639094

In [51]:
p_bruised_given_edible = p_bruised_and_edible/p_edible
p_bruised_given_edible

0.6539923954372624

In [52]:
edible = df[df['edible-poisonous'] == 'edible']
edible['bruised'].value_counts(normalize=True)

True     0.653992
False    0.346008
Name: bruised, dtype: float64

### Intuition behind conditional probability: 

How do you compute the probability that mushrooms are edible given they are bruised? 

When you ask the question "what is the probability that the mushrooms are edible and bruised?", the sample space originally contains all 8124 rows of mushrooms. 

<img src="images/Image_72_Cond4.png" width="300">

However, to compute the probability that the mushrooms are edible given they are bruised, you need to consider the reduced size of the sample space. 

In the image above, S is the universe of all mushrooms in the dataset, A is the set of mushrooms that are edible, and B is the set of mushrooms that are bruised.

* When you ask the question "what is the probability that the mushrooms are edible given the mushrooms are bruised?", you have effectively reduced the size of the sample space to include only those mushrooms that are bruised. 

* Given that mushrooms are bruised, the only way for the mushrooms to be edible is for these mushrooms to fall in the intersection of the set of mushrooms that are edible _and_ the set of mushrooms that are bruised , $P(edible \cap bruised)$.  

* To account for the smaller sample space, you divide the probability mushrooms are edible and bruised by the probability the mushrooms are bruised: $$\large P(edible|bruised) = \frac{P(edible \cap bruised)}{P(bruised)}$$




## Law of Total Probability

Sometimes we want to calculate an **unconditional** probability by making use of some **conditional** probabilities. Enter the Law of Total Probability:

$$\large P(A)=\sum _{n}P(A\cap B_{n})$$

Which works out to: 

$$\large P(A)=\sum _{n}P(A\mid B_{n}) * P(B_{n})$$

Where $B_{n}$ captures all of the disjoint events that combine to define the sample space, so $\sum P(B_{n}) = 1$.

#### In practice:

Let's say we want to know the likelihood that a mushroom is **edible**. If we know the likelihood that all tapering stalk mushrooms are edible, and the likelihood that not-tapering-stalk mushrooms are edible, plus how likely it is that mushrooms have tapering stalks (and thus also how likely the mushroom does _not_ have a tapering stalk, the compliment of the mushrooms having a tapering stalk) - well, we can use all of that to find the likelihood our mushroom will be edible! 

$$\large P(\text{edible}) = P(\text{edible} | \text{tapering}) * P(\text{tapering}) + P(\text{edible} | \text{not tapering}) * P(\text{not tapering})$$

We can do this because 'tapering stalk' and 'not tapering stalk' capture all of the possibilities of mushrooms - they define our entire sample space!

In [ ]:
# First let's find our tapering and not-tapering subsets
tapering = df.loc[df['stalk-shape'] == 'tapering']

not_tapering = df.loc[df['stalk-shape'] != 'tapering']

In [ ]:
# Calculate our probability of tapering, and probability of not-tapering
p_tapering = len(tapering) / len(df)

p_not_tapering = len(not_tapering) / len(df)
# same as 
# p_not_tapering = 1 - p_tapering

In [ ]:
# Now calculate our two conditional probabilties
p_edible_given_tapering = len(
    tapering.loc[tapering['edible-poisonous'] == 'edible']) / len(tapering)

p_edible_given_not_tapering = len(
    not_tapering.loc[not_tapering['edible-poisonous'] == 'edible']) / len(not_tapering)

In [ ]:
# Do the math!
p_edible_given_tapering * p_tapering + p_edible_given_not_tapering * p_not_tapering

In [ ]:
# Sanity check
len(df.loc[df['edible-poisonous'] == 'edible'])/len(df)

## Partitioning Complex Events

You're not really a mushroom expert, but you can see that the mushroom in your hand is has some orange on the stalk. Given the data at your disposal, what's the probability that the mushroom is edible?



$$\large P(\text{edible|orange stalk}) = \frac{P(\text{edible} \cap \text{orange stalk})}{P(\text{orange stalk})}$$

Furthermore, we can decompose $P(\text{orange stalk})$ into the two possibilities:

$P(\text{orange stalk}) = P(\text{orange stalk below ring}\cup\text{orange stalk above ring})$

But be careful here! 

$P(\text{orange stalk below ring}\cup \text{orange stalk above ring})$ **DOES NOT EQUAL** $P(\text{orange stalk below ring}) + P(\text{orange stalk above ring})$

While this may seem correct, adding these individual probabilities **double counts** mushrooms which have entirely orange stalks. However - when we do these things in Pandas (using a loc and an or condition), it won't duplicate the row - so you don't need to worry about it here!

In [ ]:
# Can prove that P(orange above) + P(orange below) != p(orange stalk)
p_orange_above = df[df['stalk-color-above-ring'] == 'orange'].shape[0]/df.shape[0]
p_orange_below = df[df['stalk-color-below-ring'] == 'orange'].shape[0]/df.shape[0]

In [ ]:
p_orange_stalk = df[(df['stalk-color-above-ring'] == 'orange')
                    | (df['stalk-color-below-ring'] == 'orange')
                    ].shape[0]/df.shape[0]

In [ ]:
p_orange_stalk == (p_orange_above + p_orange_below)

In [ ]:
# Now let's find P(edible | orange stalk)
p_edible_and_orange = df[((df['stalk-color-above-ring'] == 'orange')
                          | (df['stalk-color-below-ring'] == 'orange')
                          )
                         & (df['edible-poisonous'] == 'edible')
                         ].shape[0]/df.shape[0]

p_edible_given_orange = p_edible_and_orange / p_orange_stalk

p_edible_given_orange

# Apparently orange-stalk mushrooms seem fairly safe....
# (Disclaimer: don't take this as definitive foraging adivce!)

In [ ]:
# Sanity check...
df.loc[(df['stalk-color-above-ring'] == 'orange')
       | (df['stalk-color-below-ring'] == 'orange')]['edible-poisonous'].value_counts()

## Extra Credit: Additional Conditional Probability Practice

What's the probability that a mushroom is poisonous if it has close gill spacing AND a tapering stalk?

$$\large P(poisonous|close \cap tapering) = \frac{P(poisonous \cap close \cap tapering)}{P(close \cap tapering)}$$

In [ ]:
# P that mushroom is poisonous given close gill spacing and tapering stalk
